File: day5.exs
defmodule Aoc do
	# Breaks input into groups. Returns list of groups, where each group is a list of integers
	def parse(input) do
		input
		|> String.trim
		|> String.split("\n")
		|> Enum.map(fn row ->
			row.match(/move (?<count>\d+) from (?<from>\d+) to (?<to>\d+)/);
	end

	def part1(input) do
		input
		# |> Enum.map(fn row ->
		# 	[a1, a2, b1, b2] = String.split(row, ~r{[-,]}) |> Enum.map(fn s -> Integer.parse(s) end)
		# 	aContains = (a1 <= b1 && a2 >= b2);
		# 	bContains = (b1 <= a1 && b2 >= a2);
		# 	case aContains or bContains do
		# 		true -> 1
		# 		false -> 0
		# 	end
		# end)
		# |> Enum.sum
	end

	def part2(input) do
		parseGroups(input)
		|> Enum.map(fn row ->
			[a1, a2, b1, b2] = String.split(row, ~r{[-,]}) |> Enum.map(fn s -> Integer.parse(s) end)
			aContains = (a2 >= b1 && a1 <= b2);
			bContains = (b2 >= a1 && b1 <= a2);
			case aContains or bContains do
				true -> 1
				false -> 0
			end
		end)
		|> Enum.sum
	end

	def run(inputFile) do
		IO.puts(["Reading ", inputFile])
		{:ok, input} = File.read(inputFile)
		IO.inspect ["Part 1:", part1(parse(input))]
		# IO.inspect ["Part 2:", part2(input)]
	end
end

Aoc.run("data/sample.5")
# Aoc.run("data/input.5")

File: day5.js
/// Utility functions
const numSort = (a, b) => a - b;
const sum = (a, b) => a + b;
///

let {sample, input} = require('./input5.js');

function parse(input) {
	let stacks = [];
	let moves = [];
	let rows = input.split("\n");
	for (const row of rows) {
		let move = row.match(/move (?<count>\d+) from (?<from>\d+) to (?<to>\d+)/);
		if (move) {
			moves.push(move.groups);
		} else {
			//let stack = Array.from(row.matchAll(/(\[\a\]\s)|(\s\s\s\s)/g));
			let stackMatch = row.matchAll(/([A-Z])/g);
			for (const match of stackMatch) {
				console.log(match.index, match[0], match[1]);
				let s = (match.index-1)/4;
				if (!Array.isArray(stacks[s])) {
					stacks[s] = [];
				}
				stacks[s].push(match[0]);
			}
		}
	}

	return {
		stacks: stacks.map(s => s.reverse()),
		moves
	};
}

function part1(input) {
	let stacks = input.stacks;
	for (const move of input.moves) {
		console.log(move);
		for (let c = 0; c < parseInt(move.count); c++) {
			let from = parseInt(move.from),
				to = parseInt(move.to);
			console.log('Moving from,to: ', from, to);
			stacks[to-1].push(stacks[from-1].pop());
		}
	}
	console.log(stacks);

	return stacks.map(s => s.pop());
}

function part2(input) {
	let stacks = input.stacks;
	for (const move of input.moves) {
		console.log(move);
		let from = parseInt(move.from),
			to = parseInt(move.to),
			count = parseInt(move.count);
		console.log('Moving from,to: ', count, from, to);
		stacks[to-1] = stacks[to-1].concat(stacks[from-1].splice(-move.count));
	}
	console.log(stacks);

	return stacks.map(s => s.pop());
}

let parsed = parse(sample);
console.log(parsed);

//result = part2(parsed).join('');
//console.log(result);



